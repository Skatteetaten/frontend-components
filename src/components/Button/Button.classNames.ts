import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { getFocusStyle } from '..';
import { FontSizes, FontWeights } from '..';
import { PaletteProps } from '..';
// @ts-ignore TODO
function getTypeColor(props): string {
  const palette = getTheme().palette as PaletteProps;
  const radius = '20px';

  switch (props.type) {
    case 'primary':
      // @ts-ignore TODO
      return {
        borderRadius: '6px',
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.blue,
        color: palette.skeColor.white,
        boxShadow: `0 8px 6px -6px ${palette.skeColor.lightGrey}`,
        selectors: {
          '@media  only screen and (max-width: 479px)': {
            width: '100%',
            padding: '21px 0'
          }
        }
      };
    case 'primaryLarge':
      // @ts-ignore TODO
      return {
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.blue,
        color: palette.skeColor.white,
        boxShadow: `0 8px 6px -6px ${palette.skeColor.lightGrey}`,
        fontSize: FontSizes.largePlus,
        borderWidth: '3px',
        borderRadius: '7px',
        fontWeight: FontWeights.semibold,
        padding: '25px',
        height: '80px',
        maxWidth: 'calc((75*550px)/100)'
      };
    case 'primaryRounded':
      // @ts-ignore TODO
      return {
        borderRadius: radius,
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.white,
        color: palette.skeColor.blue
      };
    case 'primaryRoundedFilled':
      // @ts-ignore TODO
      return {
        borderRadius: radius,
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.blue,
        color: palette.skeColor.white
      };
    case 'warning':
      // @ts-ignore TODO
      return {
        borderRadius: '6px',
        borderColor: palette.skeColor.lightPink,
        background: palette.skeColor.lightPink,
        color: palette.bodyText,
        selectors: {
          '@media  only screen and (max-width: 479px)': {
            width: '100%',
            padding: '21px 0'
          }
        }
      };
    case 'secondary':
      // @ts-ignore TODO
      return {
        padding: 0,
        borderWidth: 0,
        background: 'none',
        color: palette.skeColor.blue
      };
    default:
      // @ts-ignore TODO
      return {
        borderRadius: radius,
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.white,
        color: palette.skeColor.blue
      };
  }
}
// @ts-ignore TODO
function getTypeFocusColor(props): object {
  const palette = getTheme().palette as PaletteProps;
  switch (props.type) {
    case 'warning':
      return {
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.lightPink,
        color: palette.bodyText
      };
    case 'secondary':
      return {
        background: 'none',
        color: palette.skeColor.blue
      };
    default:
      return {
        borderColor: palette.skeColor.blue,
        background: palette.skeColor.lightBlue,
        color: palette.skeColor.blue
      };
  }
}
// @ts-ignore TODO
function getTypeActiveColor(props) {
  const palette = getTheme().palette as PaletteProps;
  switch (props.type) {
    case 'warning':
      return {
        borderColor: palette.skeColor.pink,
        background: palette.skeColor.pink,
        color: palette.skeColor.white
      };
    case 'secondary':
      return {
        background: 'none',
        color: palette.skeColor.darkBlue
      };
    default:
      return {
        borderColor: palette.skeColor.darkBlue,
        background: palette.skeColor.darkBlue,
        color: palette.skeColor.white
      };
  }
}
// @ts-ignore TODO
function getDisabledColor(props) {
  switch (props.type) {
    case 'primary':
      return {
        boxShadow: `none`
      };
    case 'secondary':
      return {
        background: 'none'
      };
    default:
      return;
  }
}
// @ts-ignore TODO
function getLabelStyles(props) {
  switch (props.type) {
    case 'secondary':
      return {
        textDecoration: 'underline'
      };
    default:
      return;
  }
}
// @ts-ignore TODO
function setFocusRadius(props) {
  if (props.type === 'primary' || props.type === 'warning') {
    return '6px';
  }
  if (props.type === 'primaryLarge') {
    return '7px';
  } else {
    return '20px';
  }
}
// @ts-ignore TODO
function setInset(props) {
  if (props.type === 'primary' || props.type === 'warning') {
    return -3;
  } else {
    return -4;
  }
}
// @ts-ignore TODO
export function getClassNames(props): string {
  const theme = getTheme();
  const palette = theme.palette as PaletteProps;
  const inset = setInset(props);
  const radius = setFocusRadius(props);
  // @ts-ignore TODO
  return mergeStyles([
    getFocusStyle(theme, inset, 'relative', radius),
    {
      displayName: `ske-${props.type}-button`,
      selectors: {
        '&.ms-Button': {
          borderWidth: '3px',
          fontWeight: 'normal',
          padding: '15px',
          transition: 'background 0.3s',
          // @ts-ignore TODO
          ...getTypeColor(props)
        },
        '&.ms-Button:hover, &.ms-Button:focus': {
          ...getTypeFocusColor(props)
        },
        '&.ms-Button:hover .ms-Button-label': {
          ...getLabelStyles(props)
        },
        '&.ms-Button:active': {
          ...getTypeActiveColor(props)
        },
        '&.ms-Button:disabled': {
          background: palette.skeColor.whiteGrey,
          borderColor: palette.skeColor.lightGrey,
          color: palette.skeColor.lightGrey,
          ...getDisabledColor(props)
        },
        '&.ms-Button i': {
          fontSize: FontSizes.icon
        }
      }
    }
  ]);
}
