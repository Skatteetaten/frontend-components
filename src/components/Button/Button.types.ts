import { IButtonProps } from '@fluentui/react';

export interface ButtonProps extends IButtonProps {
  /** Ikon som skal vises foran teksten på knappen */
  icon?: string;
  /** Benyttes for å definere type knapp som skal benyttes */
  buttonStyle?:
    | 'primary'
    | 'primaryCornered'
    | 'warning'
    | 'secondary'
    | 'secondarySimple'
    | 'callToAction';
  /**
   * Skjulte props
   */
  /** @ignore Fjernet til fordel for buttonStyle */
  buttonType?: IButtonProps['buttonType'];
  /** @ignore */
  allowDisabledFocus?: IButtonProps['allowDisabledFocus'];
  /** @ignore */
  checked?: IButtonProps['checked'];
  /** @ignore */
  disabled?: IButtonProps['disabled'];
  /** @ignore */
  split?: IButtonProps['split'];
  /** @ignore */
  getSplitButtonClassNames?: IButtonProps['getSplitButtonClassNames'];
  /** @ignore */
  splitButtonAriaLabel?: IButtonProps['splitButtonAriaLabel'];
  /** @ignore */
  splitButtonMenuProps?: IButtonProps['splitButtonMenuProps'];
  /** @ignore */
  toggle?: IButtonProps['toggle'];
  /** @ignore */
  primary?: IButtonProps['primary'];
  /** @ignore */
  primaryActionButtonProps?: IButtonProps['primaryActionButtonProps'];
  /** @ignore */
  primaryDisabled?: IButtonProps['primaryDisabled'];
  /** @ignore */
  secondaryText?: IButtonProps['secondaryText'];
  /** @ignore */
  toggeled?: boolean;
}
