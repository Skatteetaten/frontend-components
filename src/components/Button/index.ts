export { Button as default } from './Button';
export * from './Button';
export type { ButtonProps } from './Button.types';
