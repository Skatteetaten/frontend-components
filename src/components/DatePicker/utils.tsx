export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
