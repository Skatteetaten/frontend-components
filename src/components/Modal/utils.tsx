import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
