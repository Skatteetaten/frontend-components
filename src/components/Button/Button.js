import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { DefaultButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import { getClassNames as getStandardClassNames } from './Button.classNames';

/**
 * @visibleName Button (Knapp)
 */
export default class Button extends React.PureComponent {
  static propTypes = {
    /** Gjør knappen til primærknapp med fylt farge */
    primary: PropTypes.bool,
    /** Benyttes for å definere type knapp som skal benyttes */
    buttonType: PropTypes.oneOf([
      'primary',
      'primaryRounded',
      'primaryRoundedFilled',
      'warning',
      'secondary',
      'primaryLarge'
    ]),
    /** Viser knappen som inaktiv */
    disabled: PropTypes.bool,
    /** Ikon som skal vises foran teksten på knappen */
    icon: PropTypes.string,
    /** Callback for klikk på knappen */
    onClick: PropTypes.func,
    className: PropTypes.string
  };
  static defaultProps = {
    primary: false,
    buttonType: 'primaryRounded',
    disabled: false,
    icon: undefined,
    onClick: undefined
  };

  render() {
    const {
      children,
      primary,
      buttonType,
      icon,
      disabled,
      onClick,
      componentRef,
      className,
      ...props
    } = this.props;
    return (
      <DefaultButton
        {...props}
        primary={primary}
        className={classnames(getStandardClassNames(this.props), className)}
        buttonType={buttonType}
        iconProps={icon && { iconName: icon }}
        disabled={disabled}
        onClick={onClick}
        componentRef={componentRef}
      >
        {children}
      </DefaultButton>
    );
  }
}
