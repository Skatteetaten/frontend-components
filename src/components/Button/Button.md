** Button brukes til hovedhandlinger, og tommelfingerregelen er det kun skal være én slik knapp på siden. **

Tydelig knapp:

```js
import Button from '@skatteetaten/frontend-components/Button';

<div className="ExampleSpacing8">
  <Button type="primary">Godkjenn</Button>{' '}
</div>;
```

Noe nedtonet hovedhandling:

```js
import Button from '@skatteetaten/frontend-components/Button';

<div className="ExampleSpacing8">
  <Button type="primaryRoundedFilled">Send inn uendret</Button>
  <Button icon="edit" type="primaryRounded">
    Endre
  </Button>
</div>;
```

Advarsel:

```js
import Button from '@skatteetaten/frontend-components/Button';

<div className="ExampleSpacing8">
  <Button type="warning">Avvis</Button>
</div>;
```

Ekstra fremtredende hovedhandling

```js
import Button from '@skatteetaten/frontend-components/Button';

<div className="ExampleSpacing8">
  <Button type="primaryLarge">Send inn meldingen</Button>
</div>;
```

```js noeditor
import Accordion from '../Accordion';
import AccordionItem from '../Accordion/AccordionItem';
<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      Button kan være firkantet eller avrundet, der den avrundede kan være fylt
      eller ikke fylt. Man skal ikke plassere avrundet og firkantet ved siden av
      hverandre. Hvilken du skal velge er avhengig av hvor mye oppmerksomhet du
      ønsker funksjonen skal ha på siden.
    </p>
    <p>
      Dersom du har en underordnet funksjon ved siden av en hovedfunksjon (for
      eksempel «Avbryt») finnes det et egen stil for dette.
    </p>
    <p>
      Advarselsknapp (Button med rød bakgrunn) brukes til en handling som er
      sidestilt med en annen primærhandling der du ønsker å signalisere en
      konskvens, typisk noe som ikke kan reverseres uten videre. (F.eks.
      Godkjenn/Avvis)
    </p>
    <p>
      Etterstreb å bruke samme type Button i løsningen for å sikre konsistens.
    </p>
    <p>
      Unngå å bruke knapper i inaktiv tilstand (disabled). I stedet bør knappen
      alltid være aktiv (blå) og, dersom skjemaet eller siden inneholder feil,
      vise fornuftige feilmeldinger når den klikkes på.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>
        Bruk én linje med tekst inne i knappen. For mye tekst kan virke
        forvirrende for skjermlesere.
      </li>
      <li>
        Ikke bruk knappen for å navigere videre til et annet område. Knappen
        skal se ut som en knapp, og ikke være en lenke.
      </li>
      <li>Ikke putt et ikon inne i selve knappen. Bruk heller IconButton.</li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/button#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
