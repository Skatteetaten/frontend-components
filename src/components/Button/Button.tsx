import classnames from 'classnames';
import {
  DefaultButton,
  IButtonProps
} from 'office-ui-fabric-react/lib-commonjs/Button';
import * as React from 'react';
import { getClassNames as getStandardClassNames } from './Button.classNames';

interface ButtonProps extends IButtonProps {
  /** Ikon som skal vises foran teksten på knappen */
  icon?: string;
  /** Benyttes for å definere type knapp som skal benyttes */
  type?:
    | 'primary'
    | 'primaryRounded'
    | 'primaryRoundedFilled'
    | 'warning'
    | 'secondary'
    | 'primaryLarge';
}
/**
 * @visibleName Button (Knapp)
 */
export default class Button extends React.PureComponent<ButtonProps, {}> {
  static defaultProps = {
    disabled: false,
    icon: undefined,
    onClick: undefined,
    primary: false,
    type: 'primaryRounded'
  };
  render() {
    const { children, icon, className, iconProps, ...props } = this.props;
    return (
      <DefaultButton
        {...props}
        className={classnames(getStandardClassNames(this.props), className)}
        iconProps={icon ? { iconName: icon } : iconProps}
      >
        {children}
      </DefaultButton>
    );
  }
}
