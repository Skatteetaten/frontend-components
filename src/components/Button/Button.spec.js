import React from 'react';
import { matches } from './../utils/test-utils';
import { mount } from 'enzyme';
import Button from './Button';

function oppsettFullDOM(props) {
  return mount(<Button {...props}>Knappetekst</Button>);
}

describe('Button komponent ', () => {
  it('matcher snapshot', () => {
    matches(<Button>Knapptekst</Button>);
  });

  it('renders Button med riktig DOM', () => {
    const wrapper = oppsettFullDOM();
    const label = wrapper.find('.ms-Button-label');

    expect(wrapper.prop('type')).toEqual('primaryRounded');
    expect(wrapper.prop('disabled')).toEqual(false);
    expect(wrapper.prop('primary')).toEqual(false);
    expect(wrapper.prop('icon')).toEqual(undefined);
    expect(label.html('icon')).toContain('Knappetekst');
  });

  it('setter Button med riktige props', () => {
    const wrapper = oppsettFullDOM({
      type: 'secondary',
      icon: 'Add',
      disabled: true
    });

    expect(wrapper.prop('type')).toEqual('secondary');
    expect(wrapper.prop('disabled')).toEqual(true);
    expect(wrapper.prop('primary')).toEqual(false);
    expect(wrapper.prop('icon')).toEqual('Add');
  });
});
