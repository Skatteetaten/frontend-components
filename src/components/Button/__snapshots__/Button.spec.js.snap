// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button komponent  matcher snapshot 1`] = `
<button
  className="ms-Button ms-Button--default ske-primaryRounded-button-41"
  data-is-focusable={true}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  type="primaryRounded"
>
  <span
    className="ms-Button-flexContainer flexContainer-42"
    data-automationid="splitbuttonprimary"
  >
    <span
      className="ms-Button-textContainer textContainer-43"
    >
      <span
        className="ms-Button-label label-45"
        id="id__0"
      >
        Knapptekst
      </span>
    </span>
  </span>
</button>
`;
